File: tests/fixtures/java/Wiring.java
// Spring-style DI wiring: interface, annotated implementation, and a
// consumer with a ctor-injected typed field used through `this.<field>`.
// Exercised by tests/java_test.rs (extracts_java_di_type_captures).
package com.example.wiring;

import java.util.List;

interface UserStore {
    void save(String user);
}

interface Lifecycle extends AutoCloseable {
}

@Component
class DbUserStore implements UserStore {
    public void save(String user) {
        System.out.println(user);
    }
}

class SignupService extends BaseService implements Lifecycle, Auditable {
    private final UserStore store;
    private List<Signup> pending;
    @Autowired
    private Mailer mailer;
    private com.acme.Notifier notifier;
    private int retries;

    SignupService(UserStore store, Mailer mailer) {
        this.store = store;
        this.mailer = mailer;
    }

    void register(String user) {
        Greeter greeter = new Greeter();
        var clock = new Clock();
        this.store.save(user);
        greeter.hello(user);
    }

    void enqueue(List<String> tags) {
        Map<String, Signup> index = new HashMap<>();
        this.mailer.send(tags);
    }
}

record AuditEvent(String kind) implements Auditable {
    public void audit() {
        System.out.println(kind);
    }
}

File: tests/fixtures/kotlin/wiring.kt
package com.example.wiring

import org.springframework.stereotype.Service

class Order(val id: String)

interface OrderStore {
    fun save(order: Order): Order
}

interface Auditable {
    fun audit(): String
}

open class BaseUseCase

class AuditTrail {
    fun record(id: String) {}
}

@Service
class JdbcOrderStore : OrderStore {
    override fun save(order: Order): Order {
        return order
    }
}

// Singleton implementation via object declaration.
object InMemoryStore : OrderStore {
    val fallback: AuditTrail = AuditTrail()

    override fun save(order: Order): Order = order
}

// Interface implementation by delegation.
class DelegatingStore(private val impl: OrderStore) : OrderStore by impl

@Service
class OrderService(
    private val store: OrderStore,
    val tracer: Tracer?,
    private val clock: java.time.Clock,
    val tags: List<String>,
    logger: Logger,
) : BaseUseCase(), Auditable {
    val audit = AuditTrail()

    fun place(order: Order, note: String?): Order {
        val trail = AuditTrail()
        trail.record(order.id)
        audit.record(order.id)
        return store.save(order)
    }

    override fun audit(): String = "orders"
}
